src/main/scala/is/hail/variant/HTSGenotypeView.scala

```diff
@@ -215,6 +206,70 @@ private class StructGenotypeView(rs: TStruct) extends HTSGenotypeView {
     val elementOffset = HTSGenotypeView.tArrayInt32.elementOffset(pxOffset, length, idx)
     m.loadInt(elementOffset)
   }
+}
 
-  def isLinearScale: Boolean = false
+object ArrayGenotypeView {
+  val tArrayFloat64 = TArray(TFloat64)
 }
+
+class ArrayGenotypeView(rowType: TStruct) {
+  private val tgs = rowType.fieldType(3).asInstanceOf[TArray]
+  private val tg = tgs.elementType match {
+    case tg: TStruct => tg
+    case _ => null
+  }
+
+  private def lookupField(name: String, expected: Type): (Boolean, Int) = {
+    if (tg != null) {
+      tg.fieldIdx.get(name) match {
+        case Some(i) => (true, i)
+        case None => (false, 0)
+      }
+    } else
+      (false, 0)
+  }
+
+  private val (gtExists, gtIndex) = lookupField("GT", TCall)
+  private val (gpExists, gpIndex) = lookupField("GP", ArrayGenotypeView.tArrayFloat64)
+
+  private var m: MemoryBuffer = _
+  private var gsOffset: Long = _
+  private var gsLength: Int = _
+  private var gOffset: Long = _
+
+  var gIsDefined: Boolean = _
+
+  def setRegion(mb: MemoryBuffer, offset: Long) {
+    this.m = mb
+    gsOffset = rowType.loadField(m, offset, 3)
+    gsLength = tgs.loadLength(m, gsOffset)
+  }
+
+  def setRegion(rv: RegionValue): Unit = setRegion(rv.region, rv.offset)
+
+  def setGenotype(idx: Int) {
+    require(idx >= 0 && idx < gsLength)
+    gIsDefined = tgs.isElementDefined(m, gsOffset, idx)
+    gOffset = tgs.loadElement(m, gsOffset, gsLength, idx)
+  }
+
+  def hasGT: Boolean = gtExists && gIsDefined && tg.isFieldDefined(m, gOffset, gtIndex)
+
+  def hasGP: Boolean = gpExists && gIsDefined && tg.isFieldDefined(m, gOffset, gpIndex)
+
+  def getGT: Int = {
+    val callOffset = tg.loadField(m, gOffset, gtIndex)
+    Call(m.loadInt(callOffset))
+  }
+
+  def getGP(idx: Int): Double = {
```

Comment:
It's a bit confusing that this is named `getGP` because GP is supposed to be the Array of probabilities, not a single probability.

Reply:
Suggestion for a new name?  I was OK with it since it takes an index and I thought that made it clear enough.